Add possibility to collect metrics from branch or pullRequest analysis

lukzeg opened the pull request:
Hi  @adisreyaj , @pedroresende

I would like to propose small extension to your plugin. I would like to extend it via adding possibility to collect metrics from pullRequest or branch analysis.

For instance if the end user would like to receive analysis from pullRequest, he might define new attribute pullRequest:
`
{
      "project": "adisreyaj_compito",
      "sonarURL": "https://sonarcloud.io",
      "pullRequest": 212,
      "auth": {
        "username": "",
        "password": ""
      }
    }
`
or branch:
`
{
      "project": "adisreyaj_compito",
      "sonarURL": "https://sonarcloud.io",
      "branch": "bugfix/FixMe",
      "auth": {
        "username": "",
        "password": ""
      }
    }
`

In case if both of attrs are added, the highest priority will have pullRequest, as mostly engineers will be interested to see potential issues which they deliver to master branch.

Please let me know if this feature looks okay for you.

Kind Regards,
Lucas

lukzeg committed:
Add possibility to collect metrics from branch or pullRequest analysis

Update README.md with information about optional arguments:
- branch
- pullRequest

pedroresende commented on src/helpers/file.helpers.ts:
```diff
@@ -19,6 +19,27 @@ const isConfigured = (config: Config) => {
+export function isPullRequestOrBranchConfigured(config: Config) {
+  let returnVal = {};
+
+  let isBranchConfigured = has(config, 'branch') && !config.branch.includes('optional-branch-name');
```
replace `let` by `const` please

lukzeg replied:
fixed.

pedroresende commented on src/helpers/file.helpers.ts:
```diff
@@ -19,6 +19,27 @@ const isConfigured = (config: Config) => {
+  let returnVal = {};
+
+  let isBranchConfigured = has(config, 'branch') && !config.branch.includes('optional-branch-name');
+  let isPullRequestConfigured = has(config, 'pullRequest') && !config.pullRequest.includes('optional-pull-request-id');
```
same as above

lukzeg replied:
fixed.

pedroresende commented on src/helpers/file.helpers.ts:
```diff
@@ -19,6 +19,27 @@ const isConfigured = (config: Config) => {
+
+  if ( isBranchConfigured || isPullRequestConfigured) {
+    if ( isBranchConfigured ) {
+      returnVal = { branch: config.branch };
```
why not return directly?

lukzeg replied:
My apologies. I'm working in another project in C and we use only one return at the end of function.  Updated

pedroresende replied:
you don't have to apologise, you did a great job, my remarks are only improvements nothing else 🎉

pedroresende commented on src/helpers/file.helpers.ts:
```diff
@@ -19,6 +19,27 @@ const isConfigured = (config: Config) => {
+
+    // Give pullRequest highest priority
+    if ( isPullRequestConfigured ) {
+      returnVal = { pullRequest: config.pullRequest };
```
why not return directly?

pedroresende commented on src/helpers/file.helpers.ts:
```diff
@@ -19,6 +19,27 @@ const isConfigured = (config: Config) => {
+}
+
+export function isPullRequestOrBranchConfigured(config: Config) {
+  let returnVal = {};
```
not needed

pedroresende commented on src/helpers/file.helpers.ts:
```diff
@@ -19,6 +19,27 @@ const isConfigured = (config: Config) => {
+      returnVal = { branch: config.branch };
+    }
+
+    // Give pullRequest highest priority
```
if you want to give highest priority you can move before the `if ( isBranchConfigured ) {` and return directly

lukzeg replied:
Hi @pedroresende ,
I refactored changes. Thanks for the tip.

lukzeg commented on src/helpers/file.helpers.ts:
```diff
@@ -19,6 +19,27 @@ const isConfigured = (config: Config) => {
+      returnVal = { pullRequest: config.pullRequest };
+    }
+  }
+  return (
```
Removed.

adisreyaj commented:
@lukzeg It's really a good enhancement to have. Here are my thoughts on the approach.

I personally feel that this should be easier for the user, i.e by a dropdown or something instead of having to update the config file every time.

The config file should be considered as a one-time setup process.

What do y'all think?

@lukzeg @pedroresende

lukzeg commented:
Hi @adisreyaj , @pedroresende ,

The idea to create settings part sound really great. I think it will be great to create some milestone with such of feature. So anyone who is interested might start to work on it.

I would like to avoid of making such of change in current PRst, just to simplify the commits and parts which are delivered. I know that might be a strange, but small parts of code are much more easier to be Code Reviewed.

What do you think about creating some milestones, with ideas of future improvements required/ request by you and other users?

lukzeg committed:
Add errorMsg when cannot reach SQ server

Add errorMsg when pullRequest or branch is not longer available at SonarQube server
